### ISSUE TITLE
session 6. concept of deep-learning

### ISSUE THREAD

**Issue by Odin-son** [Odin-son/felis-catus#5]

**Comment by Odin-son** [comment]
학습의 목적은 손실 함수의 값을 가능한 낮추는 매개변수를 찾는 것, 최적화

이전까지는 SGD(확률적 경사 하강법)을 사용, 무작정 찾는거보단 좋지만...더 똑똑한 방법을 찾아야함

**Comment by Odin-son** [comment]
SGD의 단점은 비등방성(anisotropy)함수에서는 탐색 경로가 비효율적
그래서 나온게 모멘텀, AdaGrad, Adam 이라는 방법들임

**Comment by Odin-son** [comment]
모멘텀은 속도(v)를 가져와서 SGD보다 조금 나음

**Comment by Odin-son** [comment]
AdaGrad는 학습률과 관련된 이야기,
learning-rate-decay는 학습을 진행하면서 학습률을 조금씩 줄여가는 방법, AdaGrad는 이를 발전시킨 것.
각각 매개변수에 맞춤형 줄임
<img width="189" alt="e 6 5" src="https://user-images.githubusercontent.com/33476636/98338539-ad7aaa00-204d-11eb-9593-e0400218f52b.png">
<img width="198" alt="e 6 6" src="https://user-images.githubusercontent.com/33476636/98338541-ae134080-204d-11eb-90ac-15d152b3c230.png">
h는 기울기를 제곱해서 더해줌 그리고 루트(h)로 갱신시에 나눠줌
=> 이것이 의미하는 것은 매개변수 중 가장 큰 원소는 학습률이 낮아진다는 것

**Comment by Odin-son** [comment]
Adam은 이 둘을 혼합한 것...하이퍼파라미터인 편향까지 보정해준단다

**Comment by Odin-son** [comment]
그럼 다음 스텝은 가중치에 관한 이야기인데, 
초기값을 균일하게 설정할 경우 backpropagation에서 값을 균일하게 갱신하기때문에 그렇게 하면 망함

**Comment by Odin-son** [comment]
1. 데이터가 0과 1에 치우쳐 분포하게 되면 역전파의 기울기가 점점 작아지다 사라짐 (gradient vanishing)
(e.g. 표준편차가 1인 정규분포)
2. 5층의 뉴런에 0.5에 치우친 활성화값 결과 => 기울기 소실의 문제는 없으나, 이렇게 구조를 만든 의미가 없어져버림
=> 따라서 적절히 분배하는 쪽으로 해야한다.

**Comment by Odin-son** [comment]
1. xavier 초기값 (e.g. caffe framework) ; 앞계층의 노드 갯수(n), 표준편차 1 / np.sqrt(n) 의 분포를 사용함
-> sigmoid, tanh 중앙 부근이 선형
2. ReLU 를 위한 초기값 => He 초기값(kaiming He), ReLU는 음의 영역이 0이라서 2/np.sqrt(n) 인 정규분포를 사용함...(더 넓게 분포시키기 위함)

**Comment by Odin-son** [comment]
배치 정규화는 각층이 활성화 값을 적당히 퍼뜨리도록 '강제'하는 것 : 학습 속도 개선, 초기값 의존 x, 오버피팅을 억제
- 활성화함수 앞 혹은 뒤에 삽입; Affine->Batch Norm->ReLU ...
- 미니배치 단위로 평균이 0, 분산이 1이 되도록 정규화

**Comment by Odin-son** [comment]
오버피팅은 훈련데이터에만 적응되어 그외 데이터에는 제대로 대응하지 못하는 상태
- 매개변수가 많고 표현력이 높은 모델
- 훈련 데이터가 적음 

*여기서 말하는 표현력은 활성화값의 분포를 이야기함. 분포가 적절하면 표현력이 높은거

**Comment by Odin-son** [comment]
자, 그럼 이제 가중치 감소를 해봐야겠지...
학습 과정에서 패널티를 줌으로써, 큰 가중치에 대해는 큰 패널티를 부여함
모든 가중치의 손실 함수에 1/2(lamda)W^2 을 더함, backpropagation은 (lamda)W 를 더하는 거겠지.
*정규화 항으로, L1 norm = 절대값의 합, L2 norm = 가중치의 제곱의 합의 root, L_max norm = 가장 큰 값 을 사용하는데 보통 L2 norm 씀

**Comment by Odin-son** [comment]
Dropout은 신경망이 복잡해져서 가중치 감소로 해결이 안될때,  뉴런을 임의로 삭제하면서 학습하는 방법
- 학습할땐 이렇게하고, 테스트할때는 모든 뉴런을 사용하되, '삭제안한 비율' 가중치를 줌(실제로 이 비율은 안씀)

앙상블 학습은 개별적으로 학습시킨 여러 모델의 출력을 평균을 냄
- Dropout은 매번 노드를 임의로 삭제하기에 매번 다른 모델을 학습함

**Comment by Odin-son** [comment]
하이퍼파라미터의 성능을 평가할때는 시험 데이터를 사용하면 안된다.
- 시험데이터에 오버피팅 되기 때문
- 하이퍼파라미터용 검증 데이터를 만든다
- 결국, 훈련데이터(매개변수), 검증데이터(하이퍼파라미터), 시험데이터(범용성 평가)
- 보통 훈련데이터의 20%

이러나저러나 하이퍼 파라미터는 대략적으로 지정하는게 좋음
- 가령, 10^n으로 정하는 것 ( '로그 스케일로 지정한다' )
- epoch을 작게하여, 1회 평가에 걸리는 시간을 단축하는것이 효과적임

```
weight_decay = 10 **np.random.uniform(-8,-4)
lr = 10 ** np.random.uniform(-6,-2)
```

**Comment by Odin-son** [comment]
#### 신경망 학습에 중요한 기술
##### 1. 매개변수 갱신 방법 : SGD 는 anisotropy 함수에는 맞지않음, 탐색 경로가 비효율적
- Momentum : 속도의 개념을 추가하여 갱신, SGD보다 조금 나음
- AdaGrad : 각 매개변수에 맞춤으로 학습률을 조절
- Adam : Momentum + AdaGrad , 편향까지 맞춰줌..
##### 2. 가중치의 초깃값 설정
- Xavier : sigmoid, tanh ; 1/np.sqrt(n)
- He      : ReLU ; 2/np.sqrt(n)
##### 3. 배치정규화 : 각층이 활성화 값을 적당히 퍼뜨리도록 '강제'하는 것
- 학습을 빠르게 진행할 수 있으며, 초기값에 영향을 덜 받음
- 미니배치단위로 평균이 0 분산이 1로 만듦
##### 4. 오버피팅 억제
- 가중치 감소 : 모든 가중치의 손실함수에 L2 norm, 1/2*(lamda)*(W**2) 를 더함
- Dropout : 뉴런을 임의로 삭제하며 학습
##### 5. 하이퍼파라미터
- 보통 훈련 데이터의 20%를 사용, 대략적으로 정하며 실험적으로 찾아감
